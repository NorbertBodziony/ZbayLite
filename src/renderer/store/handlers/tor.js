import Immutable from 'immutable'
import net from 'net'
import { ipcRenderer } from 'electron'

import { createAction, handleActions } from 'redux-actions'
import torSelectors from '../selectors/tor'
import { successNotification } from './utils'
import notificationsHandlers from './notifications'
import electronStore from '../../../shared/electronStore'

export const client = new net.Socket()
export const defaultTorUrlProxy = 'localhost:9050'
export const defaultTorUrlBrowser = 'localhost:9150'

export const Tor = Immutable.Record(
  {
    url: '',
    enabled: false,
    error: null,
    status: ''
  },
  'Tor'
)
export const initialState = Tor()

const setEnabled = createAction('SET_TOR_ENABLED')
const setUrl = createAction('SET_TOR_URL')
const setError = createAction('SET_TOR_ERROR')
const setStatus = createAction('SET_TOR_STATUS')

export const actions = {
  setUrl,
  setEnabled,
  setError,
  setStatus
}

const checkDeafult = () => async (dispatch, getState) => {
  const url = defaultTorUrlProxy.split(':')
  const url2 = defaultTorUrlBrowser.split(':')
  let checkedUrl
  try {
    checkedUrl = defaultTorUrlProxy
    client.connect(
      url[1],
      url[0],
      () => {
        const msg = Buffer.from('050100', 'hex')
        client.write(msg)
      }
    )
  } catch (err) {
    checkedUrl = defaultTorUrlBrowser
    client.connect(
      url2[1],
      url2[0],
      () => {
        const msg = Buffer.from('050100', 'hex')
        client.write(msg)
      }
    )
  }
  client.on('data', data => {
    client.destroy() // kill client after server's response
    if (Buffer.from(data).toString('hex') === '0500') {
      dispatch(setUrl({ url: checkedUrl }))
      dispatch(setStatus({ status: 'stable' }))
      dispatch(setError({ error: '' }))
    } else {
    }
  })
}

const checkTor = () => async (dispatch, getState) => {
  const tor = torSelectors.tor(getState())
  const url = tor.url.split(':')
  dispatch(setStatus({ status: 'loading' }))
  setTimeout(() => {
    if (torSelectors.tor(getState()).status === 'loading') {
      dispatch(setStatus({ status: 'down' }))
      dispatch(setError({ error: 'Timeout error' }))
    }
  }, 5000)
  try {
    client.connect(
      url[1],
      url[0],
      () => {
        const msg = Buffer.from('050100', 'hex')
        client.write(msg)
      }
    )
  } catch (err) {
    dispatch(setStatus({ status: 'down' }))
    dispatch(setError({ error: err.message }))
  }
  client.on('error', () => {
    dispatch(setStatus({ status: 'down' }))
    dispatch(setError({ error: 'Can not establish a connection' }))
  })
  client.on('data', data => {
    client.destroy() // kill client after server's response
    if (Buffer.from(data).toString('hex') === '0500') {
      dispatch(setStatus({ status: 'stable' }))
      dispatch(setError({ error: '' }))
    } else {
      dispatch(setStatus({ status: 'down' }))
      dispatch(setError({ error: 'Can not establish a connection' }))
    }
  })
}
export const createZcashNode = torUrl => async (dispatch, getState) => {
  electronStore.set('torEnabled', !!torUrl)
  let ipAddress
  if (torUrl && torUrl.startsWith('localhost')) {
    ipAddress = torUrl.replace('localhost', '127.0.0.1')
  } else {
    ipAddress = torUrl
  }
  ipcRenderer.send('create-node', ipAddress)
  if (torUrl) {
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        successNotification({ message: `You are using Tor proxy.` })
      )
    )
  }
}
export const epics = {
  checkTor,
  createZcashNode,
  checkDeafult
}

export const reducer = handleActions(
  {
    [setEnabled]: (state, { payload: { enabled } }) => {
      if (enabled) {
        return state.set('enabled', enabled).set('status', 'down')
      } else {
        return state
          .set('enabled', enabled)
          .set('status', 'down')
          .set('url', '')
          .set('error', '')
      }
    },
    [setUrl]: (state, { payload: { url } }) => state.set('url', url).set('status', 'down'),
    [setError]: (state, { payload: { error } }) => state.set('error', error),
    [setStatus]: (state, { payload: { status } }) => state.set('status', status)
  },
  initialState
)
export default {
  actions,
  epics,
  reducer
}
