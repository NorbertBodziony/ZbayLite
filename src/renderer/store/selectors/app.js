import { createSelector } from 'reselect'

const store = s => s

export const app = createSelector(store, state => state.get('app'))

const version = createSelector(app, a => a.version)

export default {
  version
}
