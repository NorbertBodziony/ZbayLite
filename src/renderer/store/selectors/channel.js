import { createSelector } from 'reselect'
import channelsSelector from './channels'
import identitySelectors from './identity'
import pendingMessagesSelectors from './pendingMessages'

const store = s => s

const channel = createSelector(store, state => state.get('channel'))

const data = createSelector(
  channelsSelector.data,
  channel,
  (channels, channel) => channels.find(ch => ch.get('id') === channel.id)
)

export const spentFilterValue = createSelector(channel, c => c.get('spentFilterValue', -1))

export const message = createSelector(channel, c => c.get('message'))

export const pendingMessages = createSelector(
  pendingMessagesSelectors.pendingMessages,
  channel,
  (pendingMessages, channel) => pendingMessages.filter(m => m.channelId === channel.id)
)

export const messagesMeta = createSelector(channel, c => c.messages)

const _isOwner = (address, message) => message.getIn(['sender', 'replyTo']) === address

export const messages = createSelector(
  identitySelectors.data,
  messagesMeta,
  pendingMessages,
  (identity, meta, pendingMessages) => {
    const displayableBroadcasted = meta.data.map(
      m => m.set('fromYou', _isOwner(identity.address, m))
    )
    const displayablePending = pendingMessages.map(
      m => m.get('message')
        .set('error', m.get('error'))
        .set('status', m.get('status'))
        .set('id', m.get('opId'))
        .set('fromYou', _isOwner(identity.address, m.get('message')))
    )
    return displayableBroadcasted.concat(displayablePending.values())
      .sortBy(m => m.get('createdAt'))
  })

export default {
  data,
  channel,
  spentFilterValue,
  message,
  pendingMessages,
  messages
}
