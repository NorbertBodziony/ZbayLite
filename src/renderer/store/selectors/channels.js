import { createSelector } from 'reselect'

const store = s => s

const channels = createSelector(store, state => state.get('channels'))
const data = createSelector(channels, ch => ch.data)
const loading = createSelector(channels, ch => ch.loading)
const errors = createSelector(channels, c => c.get('errors'))

export default {
  channels,
  loading,
  data,
  errors
}
