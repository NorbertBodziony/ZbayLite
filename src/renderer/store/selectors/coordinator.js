import { createSelector } from 'reselect'

const store = s => s

export const coordinator = createSelector(store, state => state.get('coordinator'))

const running = createSelector(coordinator, a => a.running)

export default {
  running
}
