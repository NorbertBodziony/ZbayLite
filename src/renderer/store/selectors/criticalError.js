import { createSelector } from 'reselect'

const store = s => s

const criticalError = createSelector(store, state => state.get('criticalError'))

const message = createSelector(criticalError, error => error.message)
const traceback = createSelector(criticalError, error => error.traceback)

export default {
  criticalError,
  message,
  traceback
}
