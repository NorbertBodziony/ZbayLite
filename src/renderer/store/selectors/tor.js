import { createSelector } from 'reselect'
const store = s => s

const tor = createSelector(
  store,
  state => state.get('tor')
)
const torEnabled = createSelector(
  tor,
  tor => tor.enabled
)

export default {
  tor,
  torEnabled
}
