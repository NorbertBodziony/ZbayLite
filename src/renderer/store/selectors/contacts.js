import { createSelector } from 'reselect'
import Immutable from 'immutable'
import identitySelectors from './identity'
import directMssagesQueueSelectors from './directMessagesQueue'
import zbayMessages from '../../zbay/messages'
import operationsSelectors from './operations'
import { operationTypes } from '../handlers/operations'
import usersSelectors from './users'
import { mergeIntoOne } from './channel'

export const Contact = Immutable.Record({
  lastSeen: null,
  key: '',
  username: '',
  address: '',
  messages: Immutable.Map(),
  newMessages: Immutable.List(),
  vaultMessages: Immutable.List(),
  offerId: null
})

const store = s => s

const contacts = createSelector(store, state => state.get('contacts'))
const contactsList = createSelector(contacts, identitySelectors.removedChannels, (contacts, removedChannels) => {
  if (removedChannels.size > 0) {
    return contacts.filter(c => c.key.length === 66 && c.offerId === null && !removedChannels.includes(c.address)).toList()
  }
  return contacts.filter(c => c.key.length === 66 && c.offerId === null).toList()
}
)

const offerList = createSelector(contacts, contacts =>
  contacts.filter(c => !!c.offerId).toList()
)
const channelsList = createSelector((contacts, identitySelectors.removedChannels), (contacts, removedChannels) => {
  if (removedChannels.size > 0) {
    return contacts.filter(c => c.key.length === 78 && c.offerId === null && !removedChannels.includes(c.address)).toList()
  }
  return contacts.filter(c => c.key.length === 78 && c.offerId === null).toList()
}
)

const directMessagesContact = address =>
  createSelector(contacts, c => c.toList().find(el => el.get('address') === address))

const contact = address =>
  createSelector(contacts, c => c.get(address, Contact()))
const messages = address =>
  createSelector(contact(address), c => c.messages.toList())
const allMessages = createSelector(contacts, c =>
  c.reduce((acc, t) => acc.merge(t.messages), Immutable.Map())
)
const getAdvertById = txid =>
  createSelector(allMessages, msgs => msgs.get(txid))
const lastSeen = address => createSelector(contact(address), c => c.lastSeen)
const username = address => createSelector(contact(address), c => c.username)
const vaultMessages = address =>
  createSelector(contact(address), c => c.vaultMessages)
const newMessages = address =>
  createSelector(contact(address), c => c.newMessages)

export const queuedMessages = address =>
  createSelector(
    directMssagesQueueSelectors.queue,
    queue =>
      queue.filter(
        m => m.recipientAddress === address && m.message.get('type') < 10
      ) //  separate offer messages and direct messages
  )

export const pendingMessages = address =>
  createSelector(operationsSelectors.operations, operations =>
    operations.filter(
      o =>
        o.type === operationTypes.pendingDirectMessage &&
        o.meta.recipientAddress === address &&
        o.meta.message.get('type') < 10 //  separate offer messages and direct messages
    )
  )

export const directMessages = (address, signerPubKey) =>
  createSelector(
    identitySelectors.data,
    usersSelectors.registeredUser(signerPubKey),
    messages(address),
    vaultMessages(address),
    (identity, registeredUser, messages, vaultMessages) => {
      const userData = registeredUser ? registeredUser.toJS() : null
      const identityAddress = identity.address
      const identityName = userData ? userData.nickname : identity.name

      const fetchedMessagesToDisplay = messages.map(msg =>
        zbayMessages.receivedToDisplayableMessage({
          message: msg,
          identityAddress,
          receiver: { replyTo: identityAddress, username: identityName }
        })
      )

      const concatedMessages = fetchedMessagesToDisplay
        .concat(vaultMessages.values())
        .sortBy(m => m.get('createdAt'))
      const merged = mergeIntoOne(concatedMessages)
      return merged
    }
  )

export default {
  contacts,
  directMessagesContact,
  queuedMessages,
  pendingMessages,
  contact,
  messages,
  directMessages,
  lastSeen,
  vaultMessages,
  username,
  newMessages,
  contactsList,
  channelsList,
  offerList,
  getAdvertById,
  allMessages
}
