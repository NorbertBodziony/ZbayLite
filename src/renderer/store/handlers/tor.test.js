import Immutable from 'immutable'

import create from '../create'
import { actions, initialState, epics, client } from './tor'
import selectors from '../selectors/tor'
import bootstrap from '../../../main/zcash/bootstrap'

jest.mock('../../vault')

describe('Tor reducer handles ', () => {
  let store = null

  beforeEach(() => {
    store = create({
      initialState: Immutable.Map({
        tor: initialState
      })
    })
    jest.clearAllMocks()
  })

  describe('actions', () => {
    it('- setEnabled', () => {
      store.dispatch(actions.setEnabled({ enabled: true }))
      expect(selectors.tor(store.getState())).toMatchSnapshot()
    })
    it('- setUrl', () => {
      store.dispatch(actions.setUrl({ url: 'testhost:9999' }))
      expect(selectors.tor(store.getState())).toMatchSnapshot()
    })
    it('- setError', () => {
      store.dispatch(actions.setError({ error: 'error' }))
      expect(selectors.tor(store.getState())).toMatchSnapshot()
    })
    it('- setStatus', () => {
      store.dispatch(actions.setStatus({ status: 'up' }))
      expect(selectors.tor(store.getState())).toMatchSnapshot()
    })
  })
  describe('epics', () => {
    it('- checkTor', async () => {
      const connectMock = jest.spyOn(client, 'connect').mockImplementation(() => {})
      await store.dispatch(epics.checkTor())
      expect(connectMock).toHaveBeenCalled()
    })
    it('- createZcashNode', async () => {
      const ensureZcashParamsMock = jest
        .spyOn(bootstrap, 'ensureZcashParams')
        .mockImplementation((platform, callback) => callback())
      const spawnZcashNodeMock = jest
        .spyOn(bootstrap, 'spawnZcashNode')
        .mockImplementation(() => ({ on: () => {} }))

      await store.dispatch(epics.createZcashNode())
      expect(ensureZcashParamsMock).toHaveBeenCalled()
      expect(spawnZcashNodeMock).toHaveBeenCalled()
    })
  })
})
