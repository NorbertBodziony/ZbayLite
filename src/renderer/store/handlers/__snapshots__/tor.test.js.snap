// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tor reducer handles  actions - setEnabled 1`] = `
Immutable.Tor {
  "url": "",
  "enabled": true,
  "error": null,
  "status": "down",
}
`;

exports[`Tor reducer handles  actions - setError 1`] = `
Immutable.Tor {
  "url": "",
  "enabled": false,
  "error": "error",
  "status": "",
}
`;

exports[`Tor reducer handles  actions - setStatus 1`] = `
Immutable.Tor {
  "url": "",
  "enabled": false,
  "error": null,
  "status": "up",
}
`;

exports[`Tor reducer handles  actions - setUrl 1`] = `
Immutable.Tor {
  "url": "testhost:9999",
  "enabled": false,
  "error": null,
  "status": "down",
}
`;
